spiffe/tls_peer.go

```diff
@@ -0,0 +1,319 @@
+package spiffe
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"net"
+	"sync"
+
+	"github.com/spiffe/go-spiffe/internal"
+	"github.com/spiffe/go-spiffe/workload"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+type TLSPeerOption func(*TLSPeer) error
+
+// WithWorkloadAPIAddr overrides the address used to reach the SPIFFE Workload
+// API. By default, the SPIFFE_ENDPOINT_SOCKET environment variable is used
+// to convey the address.
+func WithWorkloadAPIAddr(addr string) func(*TLSPeer) error {
+	return func(p *TLSPeer) error {
+		p.addr = addr
+		return nil
+	}
+}
+
+// WithLogger provides a logger to the TLSPeer
+func WithLogger(log Logger) func(*TLSPeer) error {
+	return func(p *TLSPeer) error {
+		p.log = log
+		return nil
+	}
+}
+
+type TLSPeer struct {
+	log    Logger
+	addr   string
+	client *workload.X509SVIDClient
+
+	readyOnce sync.Once
+	ready     chan struct{}
+
+	mu    sync.RWMutex
+	cert  *tls.Certificate
+	roots map[string]*x509.CertPool
+}
+
+func NewTLSPeer(opts ...TLSPeerOption) (*TLSPeer, error) {
```

Comment:
Comments on all these exported functions would be nice so we can get a good result with godoc

Reply:
:+1: